src/api/iterator-walker.ts

```diff
@@ -0,0 +1,283 @@
+  private readonly state: WalkerState;
+  private readonly joinPath: joinPath.JoinPathFunction;
+
+  constructor(root: string, options: Options) {
+    this.root = normalizePath(root, options);
+    this.state = {
+      root: isRootDirectory(this.root) ? this.root : this.root.slice(0, -1),
+      paths: [],
+      groups: [],
+      counts: new Counter(),
+      options,
+      queue: new Queue(() => {}),
+      symlinks: new Map(),
+      visited: [""].slice(0, 0),
+      controller: new Aborter(),
+      fs: options.fs || nativeFs,
+    };
+
+    this.joinPath = joinPath.build(this.root, options);
+  }
+
+  get aborted(): boolean {
+    const {
+      controller,
+      options: { signal },
+    } = this.state;
+    return controller.aborted || (signal !== undefined && signal.aborted);
+  }
+
+  private shouldPushDirectory(directoryPath: string): boolean {
+    const { options } = this.state;
+    const { includeDirs, filters } = options;
+
+    if (!includeDirs) {
+      return false;
+    }
+
+    if (filters && filters.length) {
+      return filters.every((filter) => filter(directoryPath, true));
+    }
+
+    return true;
+  }
+
+  private normalizeDirectoryPath(path: string): string {
+    const { options } = this.state;
+    const { relativePaths } = options;
+
+    if (relativePaths) {
+      return path.substring(this.root.length) || ".";
+    }
+    return path || ".";
+  }
+
+  private shouldPushFile(filePath: string): boolean {
+    const { options } = this.state;
+    const { excludeFiles, filters } = options;
+
+    if (excludeFiles) {
+      return false;
+    }
+
+    if (filters && filters.length) {
+      return filters.every((filter) => filter(filePath, false));
+    }
+
+    return true;
+  }
+
+  private async resolveSymlink(
+    symlinkPath: string
+  ): Promise<{ stat: nativeFs.Stats; resolvedPath: string } | null> {
+    const { fs, options } = this.state;
+
+    if (!options.resolveSymlinks || options.excludeSymlinks) {
+      return null;
+    }
+
+    try {
+      // TODO (43081j): probably just enforce the FSLike interface has a
+      // `promises` property, and use the normal async methods instead of
+      // promisifying
+      const resolvedPath = await promisify(fs.realpath)(symlinkPath);
+      const stat = await promisify(fs.stat)(resolvedPath);
+
+      if (
+        !stat.isDirectory() ||
+        !resolveSymlink.isRecursive(symlinkPath, resolvedPath, this.state)
+      ) {
+        return { stat, resolvedPath };
+      }
+    } catch (err) {
+      if (!options.suppressErrors) {
+        throw err;
+      }
+    }
+
+    return null;
+  }
+
+  async walkDirectory(crawlPath: string, depth: number): Promise<Dirent[]> {
+    const { state } = this;
+    const {
+      fs,
+      options: { suppressErrors },
+    } = state;
+
+    if (depth < 0) {
+      return [];
+    }
+
+    state.visited.push(crawlPath);
+
+    try {
+      const entries = await promisify(fs.readdir)(crawlPath || ".", {
```

Comment:
isn't this too expensive? it's promisifying the function each time it calls a directory